lib/Semantics/check-select-stmt.cpp

```diff
@@ -0,0 +1,398 @@
+      std::map<std::u16string, parser::CharBlock> &>::type
+  GetValHandle() {
+    return selectCaseU16StringVals_;
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, std::u32string>::value,
+      std::map<std::u32string, parser::CharBlock> &>::type
+  GetValHandle() {
+    return selectCaseU32StringVals_;
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, std::int64_t>::value,
+      std::vector<SelectCaseRangeType<std::int64_t>> &>::type
+  GetRangeValHandle() {
+    return selectCaseIntRangeVals_;
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, std::string>::value,
+      std::vector<SelectCaseRangeType<std::string>> &>::type
+  GetRangeValHandle() {
+    return selectCaseStringRangeVals_;
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, std::u16string>::value,
+      std::vector<SelectCaseRangeType<std::u16string>> &>::type
+  GetRangeValHandle() {
+    return selectCaseU16StringRangeVals_;
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, std::u32string>::value,
+      std::vector<SelectCaseRangeType<std::u32string>> &>::type
+  GetRangeValHandle() {
+    return selectCaseU32StringRangeVals_;
+  }
+
+  std::map<bool, parser::CharBlock> selectCaseLogicalVals_;
+  std::map<std::int64_t, parser::CharBlock> selectCaseIntVals_;
+  std::map<std::string, parser::CharBlock> selectCaseStringVals_;
+  std::map<std::u16string, parser::CharBlock> selectCaseU16StringVals_;
+  std::map<std::u32string, parser::CharBlock> selectCaseU32StringVals_;
+  std::vector<SelectCaseRangeType<std::int64_t>> selectCaseIntRangeVals_;
+  std::vector<SelectCaseRangeType<std::string>> selectCaseStringRangeVals_;
+  std::vector<SelectCaseRangeType<std::u16string>>
+      selectCaseU16StringRangeVals_;
+  std::vector<SelectCaseRangeType<std::u32string>>
+      selectCaseU32StringRangeVals_;
+  int kind_;
+  TypeCategory category_;
+  parser::Messages &messages_;
+};
+
+void SelectStmtChecker::Leave(
+    const parser::CaseConstruct &selectCaseConstruct) {
+  const auto &selectCaseStmt{
+      std::get<parser::Statement<parser::SelectCaseStmt>>(
+          selectCaseConstruct.t)};
+  const auto &parsedExpr{
+      std::get<parser::Scalar<parser::Expr>>(selectCaseStmt.statement.t).thing};
+  std::optional<evaluate::DynamicType> selectCaseStmtType;
+  std::optional<TypeCategory> validCaseStmtTypeCategory;
+  if (const auto *expr{GetExpr(parsedExpr)}) {
+    if (auto type{expr->GetType()}) {
+      selectCaseStmtType = type;
+    }
+  }
+  if (!selectCaseStmtType) {
+    return;
+  }
+  if ((selectCaseStmtType.value().category() != TypeCategory::Integer) &&
+      (selectCaseStmtType.value().category() != TypeCategory::Character) &&
+      (selectCaseStmtType.value().category() !=
+          TypeCategory::Logical)) {  // C1145
+    context_.Say(parsedExpr.source,
+        "SELECT CASE expression must be of type CHARACTER, INTEGER, OR LOGICAL"_err_en_US);
```

Comment:
The logic of the rest of this function would be simpler if you returned after reporting this error.

Reply:
Thank you Tim!
As you mentioned I have modified to return after reporting the error to keep the logic simple.